### PR TITLE
Increase which chiral dummy groups can be turned on

### PR DESCRIPTION
**Pull request by proteneer**
This PR hopes to address the slow down introduced by chiral restraints (without disabling them). The basic PR operates as follows:

1) Identify additional chiral restraints that may involve >2 core atom that can be turned on. This essentially proceeds via a symmetry argument.  p0_restricted -> p0_restraint -> p1_restraint -> p1_restricted -> p1_unrestricted (where the contribution from the final stage is exactly -beta^(-1) ln(2))
2) We identify all chiral 4-atom centers that can be converted effectively without staggering the angle terms, in particular, for a given cijkl center, there are 4 chiral volumes induced: `{cijk, cijl, cikl, cjkl}`, which are fully turned on at one end-state, if just one of these are present at the other end state, then we can safely turn on the all the other chiral volumes. i.e. suppose `cjkl` is present at both end-states, then it's safe to turn on `{cijk, cijl, cikl}` after all the bond terms have been turned on, without staggering the rest.

### CODE REVIEW

File: timemachine/fe/single_topology.py [hunk 1]
```diff
@@ -178,23 +185,31 @@ def setup_dummy_interactions(
         if all([a in dga for a in idxs]):
             dummy_improper_idxs.append(tuple([int(x) for x in idxs]))
             dummy_improper_params.append(params)
+
```
Comment (proteneer) [inline timemachine/fe/single_topology.py:188]: (bug fix, require that every bond term also be present at the end-state).

Comment (proteneer) [reply timemachine/fe/single_topology.py:188]: (fixed)

File: timemachine/fe/system.py [hunk 2]
```diff
@@ -55,13 +55,14 @@ def callback_fn(x):
 
 def simulate_system(U_fn, x0, num_samples=20000, steps_per_batch=500, num_workers=None, minimize=True):
     num_atoms = x0.shape[0]
+
+    seed = 2023
```
Comment (badisa) [inline timemachine/fe/system.py:59]: Unused, intended to pass to `minimize_scipy`?

File: timemachine/fe/single_topology.py [hunk 3]
```diff
@@ -178,23 +196,31 @@ def setup_dummy_interactions(
...
+    # certain configuration of chiral states are symmetrizable
+    #
+    #        all allowed geometries              |  disallowed geometry
```
Comment (badisa) [inline timemachine/fe/single_topology.py:202]: To clarify, `all allowed geometries` here means that the chiral restraint can be left on for the entire simulation?

Comment (proteneer) [reply timemachine/fe/single_topology.py:202]: they can be left on at all values of lambda (with the exception of one of the bonds ci, cj, ck being broken). Then the restraint needs to be interpolated.

File: timemachine/fe/single_topology.py [hunk 4]
```diff
@@ -178,23 +196,31 @@ def setup_dummy_interactions(
...
-        three_in_dummy_group = sum(a in dummy_group for a in (center, i, j, k)) == 3
-        one_is_root_anchor_atom = sum(a == root_anchor_atom for a in (center, i, j, k)) == 1
-        case_2 = three_in_dummy_group and one_is_root_anchor_atom
-        # TODO: maybe add more cases, if we convince ourselves they are also factorizable...
-        #    e.g. if center is root_anchor_atom, and at most one of (i, j, k) is neighbor_anchor_atom
-        #   (and maybe also if root_anchor_atom, neighbor_anchor_atom are swapped in above statement)
-        if case_1 or case_2:
-            dummy_chiral_atom_idxs.append(tuple(int(x) for x in idxs))
-            dummy_chiral_atom_params.append(params)
+        if all([a in dgc for a in (center, i, j, k)]):
```
Comment (badisa) [inline timemachine/fe/single_topology.py:212]: nit: Avoid reconstructing tuple
 
```suggestion
        if all([a in dgc for a in idxs]):
```

File: timemachine/fe/single_topology.py [hunk 5]
```diff
@@ -373,38 +405,37 @@ def setup_end_state(ff, mol_a, mol_b, core, a_to_c, b_to_c):
...
+        ci = canonicalize_bond((c, i))
+        cj = canonicalize_bond((c, j))
+        ck = canonicalize_bond((c, k))
+        if ci in canon_mol_c_bond_idxs_set and cj in canon_mol_c_bond_idxs_set and ck in canon_mol_c_bond_idxs_set:
+            all_proper_dummy_chiral_atom_idxs.append((c, i, j, k))
+            all_proper_dummy_chiral_atom_params.append(p)
+        # leave the below for debugging
+        #     print("Chiral Volume", c,i,j,k, "is okay, turning on.")
+        # else:
+        #     print("Chiral Volume", c,i,j,k, "has a disabled bond, turning off.")
```
Comment (badisa) [inline timemachine/fe/single_topology.py:435]: This seems like it might be useful to have as a warning

File: timemachine/fe/single_topology.py [hunk 6]
```diff
@@ -896,8 +926,102 @@ def __init__(self, mol_a, mol_b, core, forcefield):
...
+            assert canonicalize_bond((c, i)) in dst_bond_idxs
+            assert canonicalize_bond((c, j)) in dst_bond_idxs
+            assert canonicalize_bond((c, k)) in dst_bond_idxs
+            dst_canon_regular_idxs.add(canonicalize_chiral_atom_idxs((c, i, j, k)))
+            dst_canon_inverted_idxs.add(canonicalize_chiral_atom_idxs((c, j, i, k)))
+
+        assert len(src_canon_regular_idxs.intersection(dst_canon_inverted_idxs)) == 0
+        assert len(dst_canon_regular_idxs.intersection(src_canon_inverted_idxs)) == 0
+
+        from timemachine.fe import chiral_utils
```
Comment (jkausrelay) [inline timemachine/fe/single_topology.py:1014]: nit: move import to top level

File: timemachine/fe/single_topology.py [hunk 7]
```diff
@@ -896,8 +926,102 @@ def __init__(self, mol_a, mol_b, core, forcefield):
 
         # setup end states
         self.src_system = self._setup_end_state_src()
+
+        print(len(self.src_system.chiral_atom.potential.idxs))
```
Comment (badisa) [inline timemachine/fe/single_topology.py:930]: Remove?

File: timemachine/fe/single_topology.py [hunk 8]
```diff
@@ -896,8 +926,102 @@ def __init__(self, mol_a, mol_b, core, forcefield):
...
+                    or canonicalize_chiral_atom_idxs((c, k, i, l)) in dst_canon_regular_idxs
+                )
+                flag_3 = (
+                    canonicalize_chiral_atom_idxs((c, j, k, l)) in dst_canon_regular_idxs
+                    or canonicalize_chiral_atom_idxs((c, k, j, l)) in dst_canon_regular_idxs
+                )
+                # print("Processing chiral center", c, "with neighbors", i,j,k,l)
+
+                if not flag_0 and not flag_1 and not flag_2 and not flag_3:
+                    raise ChiralConversionError()
```
Comment (badisa) [inline timemachine/fe/single_topology.py:979]: Would be good to add messages to these errors, at least to indicate number of neighbors

File: tests/test_single_topology.py [hunk 9]
```diff
@@ -1217,3 +1246,286 @@ def test_hif2a_plot_force_constants():
...
+
+    vs_0 = st.setup_intermediate_state(0.0)
+    chiral_idxs_0 = vs_0.chiral_atom.potential.idxs
+    chiral_params_0 = vs_0.chiral_atom.params
+    assert len(chiral_idxs_0) == 4
+    assert np.sum(chiral_params_0 == DEFAULT_CHIRAL_ATOM_RESTRAINT_K) == 4
+    vs_1 = st.setup_intermediate_state(1.0)
+
+    # chiral force constants should be on for all 4 of the 7
+    # chiral terms at lambda=1
```
Comment (maxentile) [inline tests/test_single_topology.py:1453]: possibly adjust description to match assertion more closely

```suggestion
    # chiral force constants should be on for 4 of the 7
    # chiral terms at both lambda=0 and lambda=1
```

File: tests/test_single_topology.py [hunk 10]
```diff
@@ -1217,3 +1246,286 @@ def test_hif2a_plot_force_constants():
...
+    chiral_idxs_1 = vs_1.chiral_atom.potential.idxs
+    chiral_params_1 = vs_1.chiral_atom.params
+    assert len(chiral_idxs_0) == len(chiral_idxs_1)
+    assert np.sum(chiral_params_1 == DEFAULT_CHIRAL_ATOM_RESTRAINT_K) == 4
+
+
+def test_chiral_core_ring_opening():
+    # test that chiral restraints are maintained for dummy atoms when we open/close a ring,
+    # but are disabled
```
Comment (maxentile) [inline tests/test_single_topology.py:1462]: visualized this case to understand it better:

image

(at lambda = 0, there's a "real chiral center" (4 idxs) + a center involving a dummy atom (3 idxs). at lambda = 1, there's just the first set of 4 idxs)

File: timemachine/fe/single_topology.py (hunk 4 above)
Comment (proteneer) [reply timemachine/fe/single_topology.py:212]: willfix

File: timemachine/fe/single_topology.py (hunk 5 above)
Comment (proteneer) [reply timemachine/fe/single_topology.py:435]: willfix

File: timemachine/fe/single_topology.py (hunk 8 above)
Comment (proteneer) [reply timemachine/fe/single_topology.py:979]: willfix

File: timemachine/constants.py [hunk 11]
```diff
@@ -21,6 +21,9 @@
 DEFAULT_PROTEIN_FF = "amber99sbildn"
 DEFAULT_WATER_FF = "amber14/tip3p"
 
+DEFAULT_CHIRAL_ATOM_RESTRAINT_K = 1000.0
+DEFAULT_CHIRAL_BOND_RESTRAINT_K = 999.9
```
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/constants.py:25]: We don't use `DEFAULT_CHIRAL_BOND_RESTRAINT_K` at all right now and will be tested more extensively later on. Right now they're different just to make sure I don't accidentally use `DEFAULT_CHIRAL_BOND_RESTRAINT_K` in place of `DEFAULT_CHIRAL_ATOM_RESTRAINT_K`

File: timemachine/fe/single_topology.py [hunk 12]
```diff
@@ -65,10 +72,13 @@ def setup_dummy_interactions_from_ff(ff, mol, dummy_group, root_anchor_atom, nbr
     bond_params, hb = top.parameterize_harmonic_bond(ff.hb_handle.params)
     angle_params, ha = top.parameterize_harmonic_angle(ff.ha_handle.params)
     improper_params, it = top.parameterize_improper_torsion(ff.it_handle.params)
-    chiral_atom_potential, _ = top.setup_chiral_restraints(chiral_atom_k)
+    chiral_atom_potential, _ = top.setup_chiral_restraints(chiral_atom_k, chiral_bond_k)
     chiral_atom_idxs = chiral_atom_potential.potential.idxs
     chiral_atom_params = chiral_atom_potential.params
 
+    # note: core atoms are not simplet set(mol_a_atoms).difference(dummy_group)
```
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/fe/single_topology.py:79]: willfix

File: timemachine/fe/single_topology.py [hunk 13]
```diff
@@ -178,23 +196,31 @@ def setup_dummy_interactions(
         if all([a in dga for a in idxs]):
             dummy_improper_idxs.append(tuple([int(x) for x in idxs]))
             dummy_improper_params.append(params)
+
+    # certain configuration of chiral states are symmetrizable
+    #
```
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/fe/single_topology.py:201]: will document, but . means dummy bond, and the right most diagram is disallowed because this function sets up dummy interactions, and a moeity with only core atoms is not a dummy interaction (by definition).

File: timemachine/fe/single_topology.py [hunk 14]
```diff
@@ -896,8 +926,102 @@ def __init__(self, mol_a, mol_b, core, forcefield):
...
+                    raise ChiralConversionError()
+
+            if len(nbs) == 3:
+                i, j, k = nbs
+                flag_0 = (
+                    canonicalize_chiral_atom_idxs((c, i, j, k)) in dst_canon_regular_idxs
+                    or canonicalize_chiral_atom_idxs((c, j, i, k)) in dst_canon_regular_idxs
+                )
+                if not flag_0:
+                    raise ChiralConversionError()
```
Comment (maxentile) [inline timemachine/fe/single_topology.py]: doc: add 1-line comment to describe how the specification of this function this differs from existing functionality in chiral_utils. (at first glance, looks like this intent could also be achieved using calls to `ChiralRestrIdxSet` instance methods, but I am probably missing a subtlety)

test: possibly add unit tests for `check_chiral_validity`, `assert_chiral_consistency_and_validity`

File: timemachine/fe/single_topology.py [hunk 15]
```diff
@@ -325,11 +351,15 @@ def setup_end_state(ff, mol_a, mol_b, core, a_to_c, b_to_c):
     # gotta add 'em all!
 
     for anchor, (nbr, dg) in dgs.items():
```
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/fe/single_topology.py:353]: willfix

File: timemachine/fe/single_topology.py [hunk 16]
```diff
@@ -1165,15 +1289,14 @@ def setup_intermediate_state(self, lamb) -> VacuumSystem:
...
 
         # stagger the lambda schedule
-        bonds_min, bonds_max = [0.0, 0.3]
-        angles_min, angles_max = [0.6, 0.8]
-        torsions_min, torsions_max = [0.8, 1.0]
-        chiral_atoms_min, chiral_atoms_max = [0.3, 0.6]
+        bonds_min, bonds_max = [0.0, 0.7]
+        angles_min, angles_max = [0.0, 0.7]
+        torsions_min, torsions_max = [0.7, 1.0]
+        chiral_atoms_min, chiral_atoms_max = [0.7, 1.0]
```
Comment (maxentile) [inline timemachine/fe/single_topology.py:1294]: Nice: intuitively, this seems more efficient.

(if these transformations no longer conflict with each other very much, seems more economical to schedule them simultaneously)

**Review by maxentile — APPROVED** [review]
Nice -- looks like a significant improvement!

(As before https://github.com/proteneer/timemachine/pull/1255#pullrequestreview-1875167054 , I cannot confidently estimate the quantity or importance of corner cases, and will mostly rely on the results of numerical tests, which may complete after merging.)

File: tests/test_chiral_restraints.py [hunk 17]
```diff
@@ -289,7 +293,8 @@ def test_chiral_restraints_tetrahedral():
     U_fn = system.get_U_fn()
 
     vols_orig = []
-    frames = simulate_system(U_fn, x0)
+    # turn off minimizing to avoid accidentally swapping chiral states
+    frames = simulate_system(U_fn, x0, minimize=False)
```
Comment (badisa) [inline tests/test_chiral_restraints.py:297]: Can you comment on the lack of minimization? Shouldn't the `U_fn` include the chiral restraints?

File: tests/test_single_topology.py [hunk 18]
```diff
@@ -1217,3 +1246,286 @@ def test_hif2a_plot_force_constants():
...
+
+@pytest.mark.nocuda
+def test_hif2a_pairs_setup_st():
+    """
+    Test that we can setup all-pairs single topology objects in hif2a.
+    """
+    with resources.path("timemachine.testsystems.data", "ligands_40.sdf") as path_to_ligand:
+        mols = read_sdf(path_to_ligand)
+
+    pairs = [(mol_a, mol_b) for mol_a in mols for mol_b in mols]
```
Comment (badisa) [inline tests/test_single_topology.py:1259]: nit: This does N^2, including self-self transformations, not sue this is informative. Might be better to do upper triangular of the pairs

File: timemachine/fe/system.py [hunk 19]
```diff
@@ -55,13 +55,14 @@ def callback_fn(x):
...
+
     if minimize:
         x_min = minimize_scipy(U_fn, x0)
```
Comment (badisa) [inline timemachine/fe/system.py:62]: nit: Should seed be passed here?

File: tests/test_single_topology.py (hunk 18 above)
Comment (proteneer) [reply tests/test_single_topology.py:1259]: intentional - I wanted to test setting up the single topology both ways `SingleTopology(mol_a, mol_b,...)` and `SingleTopology(mol_b, mol_a, ...)`

File: tests/test_single_topology.py (hunk 9 above)
Comment (proteneer) [reply tests/test_single_topology.py:1453]: willfix

**Review by badisa — APPROVED** [review]

**Review by jkausrelay — APPROVED** [review]

File: timemachine/fe/system.py (hunk 19 above)
Comment (proteneer) [reply timemachine/fe/system.py:62]: yes, will fix

File: timemachine/fe/single_topology.py (hunk 14 above)
Comment (proteneer) [reply timemachine/fe/single_topology.py]: addressed in [a82](https://github.com/proteneer/timemachine/pull/1256/commits/a82bab4b9489754abd7e2944379a176ad6a5e321)

File: tests/test_single_topology.py (hunk 10 above)
Comment (proteneer) [reply tests/test_single_topology.py:1461]: The above picture is a bit misleading since it's missing the dummy H (atomic number 6) from the right hand side.

In total, at both lambdas, there's 4 chiral idxs associated with the carbon:

`[0, 1, 4, 5], [0, 1, 4, 2], [0, 1, 5, 2], [0, 2, 4, 5]`

And 3 chiral_idxs associated with the dummy hydrogen:

`[0,6,4,5], [0, 2, 5, 6], [0, 4,6,2]`

At lambda=0, all 7 chiral terms are turned on.

At lambda=1, the 0,1 C-O bond is broken, disabling 3 of the chiral idxs:

`[0, 1, 4, 5], [0, 1, 4, 2], [0, 1, 5, 2]`

Leaving us with:

`[0, 2, 4, 5], [0,6,4,5], [0, 2, 5, 6], [0, 4,6,2]`

**Comment by proteneer** [comment]
Note to self - add one more test case morphing methyl into nitrile (as opposed to water), and ensuring that all chiral volumes are turned off.

**Comment by proteneer** [comment]
Interesting - looks like we don't need to check for the angle terms at all! The normalized chiral volume is already numerically stable even if angles terms are at 180, or disabled all together. It's only sensitive to the length but not the angle. i.e. the chiral volume defined by `(c, i, j, k)` does not change in a numerically unstable way (wrt. small perturbations on individual atoms `c,i,j,k`)  as the angle term `(i,c,j), (i,c,k), (j,c,k)` goes to zero or 180.
